From: Mike Hommey <mh+mozilla@glandium.org>
Date: Fri, 15 Jul 2011 09:35:44 +0200
Subject: Bug 671804 - Provide NS_APP_PREF_DEFAULTS_50_DIR and
 NS_APP_PREFS_DEFAULTS_DIR_LIST in xpcshell dir provider

---
 js/src/xpconnect/shell/xpcshell.cpp |   21 +++++++++++++++++++++
 1 files changed, 21 insertions(+), 0 deletions(-)

diff --git a/js/src/xpconnect/shell/xpcshell.cpp b/js/src/xpconnect/shell/xpcshell.cpp
index 2753f20..adda900 100644
--- a/js/src/xpconnect/shell/xpcshell.cpp
+++ b/js/src/xpconnect/shell/xpcshell.cpp
@@ -2093,6 +2093,15 @@ XPCShellDirProvider::GetFile(const char *prop, PRBool *persistent,
         *persistent = PR_TRUE;
         NS_ADDREF(*result = mGREDir);
         return NS_OK;
+    } else if (mGREDir && !strcmp(prop, NS_APP_PREF_DEFAULTS_50_DIR)) {
+        nsCOMPtr<nsIFile> file;
+        *persistent = PR_TRUE;
+        if (NS_FAILED(mGREDir->Clone(getter_AddRefs(file))) ||
+            NS_FAILED(file->AppendNative(NS_LITERAL_CSTRING("defaults"))) ||
+            NS_FAILED(file->AppendNative(NS_LITERAL_CSTRING("pref"))))
+            return NS_ERROR_FAILURE;
+        NS_ADDREF(*result = file);
+        return NS_OK;
     }
 
     return NS_ERROR_FAILURE;
@@ -2115,6 +2124,18 @@ XPCShellDirProvider::GetFiles(const char *prop, nsISimpleEnumerator* *result)
             dirs.AppendObject(file);
 
         return NS_NewArrayEnumerator(result, dirs);
+    } else if (!strcmp(prop, NS_APP_PREFS_DEFAULTS_DIR_LIST)) {
+        nsCOMArray<nsIFile> dirs;
+
+        nsCOMPtr<nsIFile> file;
+        if (NS_FAILED(NS_GetSpecialDirectory(NS_XPCOM_CURRENT_PROCESS_DIR,
+                                             getter_AddRefs(file))) ||
+            NS_FAILED(file->AppendNative(NS_LITERAL_CSTRING("defaults"))) ||
+            NS_FAILED(file->AppendNative(NS_LITERAL_CSTRING("preferences"))))
+            return NS_ERROR_FAILURE;
+
+        dirs.AppendObject(file);
+        return NS_NewArrayEnumerator(result, dirs);
     }
     return NS_ERROR_FAILURE;
 }
